newa/__init__.py

```diff
@@ -477,9 +479,93 @@ class Request(Cloneable, Serializable):
     id: str
     context: RecipeContext = field(factory=dict)
     environment: RecipeEnvironment = field(factory=dict)
+    compose: Optional[str] = None
     git_url: Optional[str] = None
     git_ref: Optional[str] = None
     tmt_path: Optional[str] = None
+    plan: Optional[str] = None
+
+    def fetch_details(self) -> None:
+        raise NotImplementedError
+
+    def generate_tf_exec_command(self) -> str:
+        cmdline_start = "NO_COLOR=yes testing-farm request --no-wait"
+        # check compose
+        if not self.compose:
+            print('ERROR: compose is not specified for the request')
+            return ''
+        tf_request_params = f'--compose "{self.compose}"'
+        # process git_url
+        if not self.git_url:
+            print('ERROR: git_url is not specified for the request')
+            return ''
+        tf_request_params += f' --git-url "{self.git_url}"'
+        # process git_ref
+        if self.git_ref:
+            tf_request_params += f' --git-ref "{self.git_ref}"'
+        # process tmt_path
+        if self.tmt_path:
+            tf_request_params += f' --path "{self.tmt_path}"'
+        # process plan
+        if self.plan:
+            tf_request_params += f' --plan "{self.plan}"'
+
+        # process context
+        context_params = ''
+        if self.context:
+            for k, v in self.context.items():
+                context_params += f' -c {k}="{v}"'
+        # process environment
+        environment_params = ''
+        if self.environment:
+            for k, v in self.environment.items():
+                environment_params += f' -e {k}="{v}"'
+
+        return f'{cmdline_start} {context_params} {environment_params} {tf_request_params}'
+
+    def initiate_tf_request(self) -> TFRequest:
+        command = self.generate_tf_exec_command()
+        if not command:
+            raise Exception("Failed to generate testing-farm command")
+        try:
+            process = subprocess.run(
+                command,
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                text=True,
+                shell=True)
+            output = process.stdout
+        except subprocess.CalledProcessError as e:
+            output = e.stdout
+        r = re.search(' api [^h]*(https://[\\S]*)\x1b', output)
+        if not r:
+            raise Exception(f"TF request failed:\n{output}\n")
+        api = r.group(1).strip()
+        request_uuid = api.split('/')[-1]
+        return TFRequest(api=api, uuid=request_uuid)
+
+
+@define
+class TFRequest(Cloneable, Serializable):
+    """ A class representing plain Testing Farm request """
+
+    api: str
+    uuid: str
+    details: Optional[dict[str, Any]] = None
```

Comment:
@happz Would you have suggestion what should be the type here? I wanted to use `JSON` as similarly to `ErrataTool.fetch_info()` it would return JSON but when using `JSON` as type I was getting an error that it is not indexable.

Reply:
Hm, I wonder where that complain comes from, because `details: Optional[JSON] = None` should be fine.

Reply:
Forget it. It does not complain now, I think it got somehow fixed by the most recent changes I did, which was moving some code from `cli.py` to classess in `__init__.py`.